src/account/account.service.ts

```diff
@@ -0,0 +1,54 @@
+import { Injectable } from "@nestjs/common";
+import { utils } from "ethers";
+import { getRPCProvider } from "../utils/client";
+import { Currency } from "../models/Currency";
+import { Errors } from "../models/Errors";
+import { PartialBlockIdentifier } from "../models/PartialBlockIdentifier";
+
+@Injectable()
+export class AccountService {
+
+  public async getBalance(accountAddress: string, blockNumber?: number) {
+    const provider = getRPCProvider()
+    let blockIdentifier = new PartialBlockIdentifier()
+    let balances = []
+    
+    if (blockNumber){      
+      // historical balance lookup            
+      const block = await provider.getBlock(blockNumber)      
+      
+      if (!block || (block.number !== blockNumber)) {        
+        return Errors.BLOCK_NOT_FOUND;  
+      }
+      blockIdentifier.index = block.number
+      blockIdentifier.hash = block.hash
+
+      const balance = await provider.getBalance(accountAddress, blockIdentifier.index)      
+      balances.push({
+        value: utils.formatEther(balance),
+        currency: Currency.EWT        
+      })
+
+    } else {    
+      // latest balance loookup
+      const blockNumber = await provider.getBlockNumber()      
+      
+      const block = await provider.getBlock(blockNumber)
+      blockIdentifier.index = block.number
+      blockIdentifier.hash = block.hash
+
+      const balance = await provider.getBalance(accountAddress)                
```

Comment:
This actually will not return consistent results, these calls are not atomic, so between executing `getBlockNumber()` and `getBalance()` the block change can happen.

I suggest to simply this whole function as follows:

```
const block = await provider.getBlock(blockNumber ?? "latest");
const balance = await provider.getBalance(accountAddress, block.number);

return {
"block_identifier": new BlockIdentifier(block.number, block.hash),
"balances": [{value: balance, currency: Currency.EWT}]
}
```

Reply:
Yes, I missed that when changed from iterative async